pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,173 @@
+// Otherwise, perform cleanup based on a combination of the OpenShift managed tag and the latest services ConfigMap.
+func Deconfigure(c client.Client, addrs []net.Addr, svcMgr manager.Manager, preserveNode bool) error {
+	var err error
+	var node *core.Node
+	ctx := context.TODO()
+	cm := &core.ConfigMap{}
+	// Fetch the CM of the desired version
+	err = func() error {
+		node, err = controller.GetAssociatedNode(c, addrs)
+		if err != nil {
+			return err
+		}
+		desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+		if !present {
+			return errors.Wrapf(err, "node missing desired version annotation")
+		}
+		return c.Get(ctx,
+			client.ObjectKey{Namespace: controller.WMCONamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	}()
+	if err == nil {
+		// Happy path, no issues getting the desired services ConfigMap
+		if err = removeCMServices(c, ctx, svcMgr, cm); err != nil {
+			return err
+		}
+	} else {
+		klog.Warningf(err.Error())
+		if err = outOfSyncCleanup(c, ctx, svcMgr); err != nil {
+			return err
+		}
+	}
+	// Delete node if needed
+	if preserveNode || node == nil {
+		return nil
+	}
+	klog.Infof("deleting node %s", node.GetName())
+	return c.Delete(ctx, node)
+}
+
+// removeCMServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the given flag is false
+func removeCMServices(c client.Client, ctx context.Context, svcMgr manager.Manager, cm *core.ConfigMap) error {
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	// Build up log message and failures
+	servicesRemoved := []string{}
+	var failedRemovals error
+	// The services are ordered by increasing priority already, so stop them in reverse order to avoid dependency issues
+	for i := len(cmData.Services) - 1; i >= 0; i-- {
+		service := cmData.Services[i]
+		if err := svcMgr.DeleteService(service.Name); err != nil {
+			failedRemovals = errors.Errorf("unable to delete %s (%v); %s", service.Name, err, failedRemovals)
+		}
+		servicesRemoved = append(servicesRemoved, service.Name)
+	}
+	if failedRemovals != nil {
+		return failedRemovals
+	}
+	klog.Infof("deleted services %v from ConfigMap %s", servicesRemoved, cm.Name)
+	return nil
+}
+
+// outOfSyncCleanup removes all services on the instance with the Openshift managed tag, as well as any services defined
+// in the latest retreivable version of the services ConfigMap
+func outOfSyncCleanup(c client.Client, ctx context.Context, svcMgr manager.Manager) error {
+	klog.Infof("cleaning up all %s Windows services", windows.ManagedTag)
+	allServices, err := svcMgr.GetServices()
+	if err != nil {
+		return errors.Wrap(err, "could not determine existing Windows services")
+	}
+	// Build up any failures
+	var failedRemovals error
+	for service := range allServices {
+		// Re-query existing services to avoid processing one that has already been removed
+		currServices, err := svcMgr.GetServices()
```

Comment:
its not clear what `currServices` means, `currentServices`? I dont think the brevity is worth the loss of clarity